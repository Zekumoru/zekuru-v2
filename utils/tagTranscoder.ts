type TTagTable = Map<string, string>;

interface ITagTranscoder {
  encode: (message: string) => [string, TTagTable];
  decode: (message: string, tagTable: TTagTable) => string;
}

// Ignore syntax: (discord tags | links | code blocks)
const DISCORD_IGNORE_REGEX =
  /(<a?[:@][\w:]*>)|(https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&\/=]*))|(```((?!\s)|\r\n|\r|\n|.)*```)/g;

const tagTranscoder: ITagTranscoder = {
  encode: (message) => {
    const matches = message.matchAll(DISCORD_IGNORE_REGEX);
    const tagTable: TTagTable = new Map();
    const tokens: string[] = [];

    let counter = 0;
    let lastIndex = 0;
    for (const match of matches) {
      // push message parts (or tokens)
      const token = message.substring(lastIndex, match.index);
      tokens.push(token);
      lastIndex = match.index + match[0].length;

      // push encoded
      const key = `<:${counter}:>`;
      tokens.push(key);
      tagTable.set(key, match[0]);
      counter++;
    }
    tokens.push(message.substring(lastIndex));

    return [tokens.join(''), tagTable];
  },
  decode: (message, tagTable) => {
    const matches = message.matchAll(DISCORD_IGNORE_REGEX);
    const tokens: string[] = [];

    let lastIndex = 0;
    for (const match of matches) {
      // push message parts (or tokens)
      const token = message.substring(lastIndex, match.index);
      tokens.push(token);
      lastIndex = match.index + match[0].length;

      // push decoded
      const key = match[0];
      tokens.push(tagTable.get(key) ?? '');
    }
    tokens.push(message.substring(lastIndex));

    return tokens.join('');
  },
};

export default tagTranscoder;
