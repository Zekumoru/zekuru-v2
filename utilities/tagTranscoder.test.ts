import tagTranscoder from './tagTranscoder';

describe('tagTranscoder', () => {
  test.each([
    [`this is an example sentence`],
    [`<@123> did you check the <#456> channel`],
    [`<:emoji:111> <:emoji:111> <:emoji:111>`],
    [`Yes sure! <:emoji:111> <a:emoji_anim:222>`],
    [`Why don't you check the official docs here: https://discord.js.org/ ?`],
    [
      `
Check the code below:
\`\`\`ts
// some code...
\`\`\`

Or you can also try:
\`\`\`ts
// some code...
\`\`\`

Hopefully that answers your question! <a:emoji_anim:333>
`,
    ],
    [`Some <@123> random <#456> text <@789>`],
  ])(`%s - that encoding/decoding works`, (text) => {
    const [encoded, tagTable] = tagTranscoder.encode(text);

    expect(encoded).toMatchSnapshot();
    expect(tagTranscoder.decode(encoded, tagTable)).toBe(text);
  });
});
