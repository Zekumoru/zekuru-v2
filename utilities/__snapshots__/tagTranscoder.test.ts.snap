// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tagTranscoder 
Check the code below:
\`\`\`ts
// some code...
\`\`\`

Or you can also try:
\`\`\`ts
// some code...
\`\`\`

Hopefully that answers your question! <a:emoji_anim:333>
 - that encoding/decoding works 1`] = `
"
Check the code below:
<:0:>

Hopefully that answers your question! <:1:>
"
`;

exports[`tagTranscoder <:emoji:111> <:emoji:111> <:emoji:111> - that encoding/decoding works 1`] = `"<:0:> <:1:> <:2:>"`;

exports[`tagTranscoder <@123> did you check the <#456> channel - that encoding/decoding works 1`] = `"<:0:> did you check the <:1:> channel"`;

exports[`tagTranscoder Some <@123> random <#456> text <@789> - that encoding/decoding works 1`] = `"Some <:0:> random <:1:> text <:2:>"`;

exports[`tagTranscoder Why don't you check the official docs here: https://discord.js.org/ ? - that encoding/decoding works 1`] = `"Why don't you check the official docs here: <:0:> ?"`;

exports[`tagTranscoder Yes sure! <:emoji:111> <a:emoji_anim:222> - that encoding/decoding works 1`] = `"Yes sure! <:0:> <:1:>"`;

exports[`tagTranscoder this is an example sentence - that encoding/decoding works 1`] = `"this is an example sentence"`;
